zcash_client_sqlite/src/wallet/init/migrations/sapling_memo_consistency.rs

```diff
@@ -0,0 +1,218 @@
+}
+
+impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &rusqlite::Transaction) -> Result<(), Self::Error> {
+        let mut stmt_raw_tx = transaction.prepare(
+            "SELECT DISTINCT sent_notes.tx, sent_notes.output_index, accounts.account, accounts.ufvk
+             FROM sent_notes 
+             JOIN accounts ON sent_notes.from_account = accounts.account
+             JOIN transactions ON transactions.id_tx = sent_notes.tx
+             WHERE transactions.raw IS NOT NULL"
+        )?;
+
+        let mut rows = stmt_raw_tx.query([])?;
+
+        let mut tx_sent_notes: BTreeMap<i64, Vec<(u32, AccountId, UnifiedFullViewingKey)>> =
+            BTreeMap::new();
+        while let Some(row) = rows.next()? {
+            let id_tx: i64 = row.get(0)?;
+            let output_index: u32 = row.get(1)?;
+            let account: u32 = row.get(2)?;
+            let ufvk_str: String = row.get(3)?;
+            let ufvk = UnifiedFullViewingKey::decode(&self.params, &ufvk_str).map_err(|e| {
+                WalletMigrationError::CorruptedData(format!(
+                    "Could not decode unified full viewing key for account {}: {:?}",
+                    account, e
+                ))
+            })?;
+
+            tx_sent_notes
+                .entry(id_tx)
+                .and_modify(|v| v.push((output_index, AccountId::from(account), ufvk.clone())))
+                .or_insert_with(|| vec![(output_index, AccountId::from(account), ufvk)]);
+        }
+
+        for (id_tx, sapling_outputs) in tx_sent_notes {
+            let (block_height, tx) =
+                get_transaction(transaction, &self.params, id_tx).map_err(|err| match err {
+                    SqliteClientError::CorruptedData(msg) => {
+                        WalletMigrationError::CorruptedData(msg)
+                    }
+                    SqliteClientError::DbError(err) => WalletMigrationError::DbError(err),
+                    other => WalletMigrationError::CorruptedData(format!(
+                        "An error was encountered decoding transaction data: {:?}",
+                        other
+                    )),
+                })?;
+
+            let ufvks: HashMap<AccountId, UnifiedFullViewingKey> = sapling_outputs
+                .iter()
+                .map(|(_, account, ufvk)| (*account, ufvk.clone()))
+                .collect();
+
+            let decrypted_outputs = decrypt_transaction(&self.params, block_height, &tx, &ufvks);
+            let mut stmt_update_sent_memo = transaction.prepare(
+                "UPDATE sent_notes 
+                SET memo = :memo 
+                WHERE tx = :id_tx
+                AND output_index = :output_index",
+            )?;
+
+            for d_out in decrypted_outputs {
+                if sapling_outputs
+                    .iter()
+                    .any(|(idx, _, _)| &(u32::try_from(d_out.index).unwrap()) == idx)
```

Comment:
When could this be false? We looked up this transaction via the `sent_notes` table, so if any change had been made to the `transaction` row that caused it to conflict with the `sent_notes` table, that seems like it would be a bug that we should raise a corruption error for instead of ignoring.

Reply:
I... don't remember what I was thinking here. The update statement also skips anything where the decrypted note doesn't have an entry in the `sent_notes` table, so this is entirely unnecessary.